src/reporter.ts

```diff
@@ -0,0 +1,290 @@
+      test.status = testStatus as SauceStatus;
+      test.output = testStep.result?.message;
+      test.duration = this.getDuration(testStep.result?.duration);
+      test.attachments = [];
+      testStep.attachments.forEach((attachment: any) => {
+        this.assets.push({
+          filename: attachment.testCaseStartedId + '.log',
+          data: Buffer.from(attachment.body),
+        })
+        test.attachments?.push({
+          name: attachment.testCaseStartedId + '.log',
+          contentType: attachment.mediaType,
+          path: ''
+        })
+      })
+      curr.addTest(test);
+    })
+    suite.addSuite(curr)
+    this.testRun.addSuite(suite);
+  }
+
+  async logTestRun(testRunFinished: { success: any; }) {
+    this.passed = testRunFinished.success;
+    this.reportToFile(this.testRun)
+
+    const id = await this.reportToSauce(this.testRun);
+    this.consoleLog.push(testRunFinished.success ? 'SUCCESS' : 'FAILURE')
+    this.logSauceJob(id as string)
+    this.log('\n')
+  }
+
+  getDuration(duration: any) {
+    return (duration.seconds * 1000) + (duration.nanos / 1000000000);
+  }
+
+  logSauceJob (jobId: string) {
+    if (jobId.length < 1) {
+      let msg = '';
+      const hasCredentials = process.env.SAUCE_USERNAME && process.env.SAUCE_USERNAME !== '' && process.env.SAUCE_ACCESS_KEY && process.env.SAUCE_ACCESS_KEY !== '';
+      if (!hasCredentials) {
+        msg = `\nNo results reported to Sauce. $SAUCE_USERNAME and $SAUCE_ACCESS_KEY environment variables must be defined in order for reports to be uploaded to Sauce.`;
+      }
+      this.log(msg);
+      return;
+    }
+    const jobUrl = this.getJobUrl(jobId as string, this.region)
+    this.log(`\nReported jobs to Sauce Labs:\n`);
+    this.log(jobUrl)
+  }
+
+  reportToFile(report: TestRun) {
+    if (!this.outputFile) {
+      return;
+    }
+
+    fs.mkdirSync(path.dirname(this.outputFile), { recursive: true });
+    report.toFile(this.outputFile);
+  }
+
+  async reportToSauce(report: TestRun) : Promise<string | undefined> {
+    // Currently no reliable way to get the browser version
+    const browserVersion = '1.0';
+
+    const jobBody = this.createBody({
+      build: this.build,
+      startedAt: this.startedAt ? this.startedAt.toISOString() : new Date().toISOString(),
+      endedAt: this.endedAt ? this.endedAt.toISOString() : new Date().toISOString(),
+      success: this.passed,
+      suiteName: this.suiteName,
+      tags: this.tags,
+      browserName: this.browserName,
+      browserVersion,
+      cucumberVersion: this.cucumberVersion, 
+    });
+
+    if (this.shouldUpload) {
+      const sessionID = await this.createJob(jobBody);
+      if (sessionID) {
+        await this.uploadAssets(sessionID);
+      }
+      return sessionID;
+    }
+  }
+
+  async uploadAssets (sessionId: string) {
+    this.assets.push({
+      filename: 'console.log',
+      data: Buffer.from(this.consoleLog.join('\n').toString()),
```

Comment:
One thought here is that the `console.log` that's being uploaded isn't actually what is printed to the console locally. Locally I see:

```🍖 npx cucumber-js --config tests/integration/cucumber.json --format ./src/reporter.ts --publish-quiet
1 scenario (1 passed)
2 steps (2 passed)
0m00.011s (executing steps: 0m00.000s)

Reported jobs to Sauce Labs:
https://app.saucelabs.com/tests/ed387ed892c042ab86b7ef13032037ba

```

But the attached console.log in the ui is:

```
Say hello	#tests/integration/features/greetings.feature
  When the greeter says hello - PASSED
  Then I should have heard "hello" - PASSED
```

So we're not actually preserving the console output. Not sure how important that is though.

Reply:
I've tried to find how to capture the log from cucumber. Haven't found a way to export that :-(

Reply:
`log` is just a member of the Formatter class. You *could* re-assign it and write to multiple streams:

```  typescript
  baseLog: (buffer: string | Uint8Array) => void;

  logWrapper(buffer: string | Uint8Array) {
    this.baseLog(buffer);
    this.consoleLog.push(Buffer.from(buffer).toString('utf8'));
  }

  constructor(config: IFormatterOptions) {
    super(config)

    this.baseLog = this.log;
    this.log = this.logWrapper;
...
```

I *think* this is safe to do?

Could also do it in a future PR. Whatever you think is best.

Reply:
It works! Thanks for the suggestion.